test/functional/api/v2/compute_resources_controller_test.rb

```diff
@@ -1,244 +1,274 @@
+    put :update, { :id => compute_resources(:yourcompute).to_param, :compute_resource => { :description => "new_description" } }
+    assert_response :not_found
+  end
+
+  test "should not destroy compute resource for restricted" do
+    setup_user 'destroy', 'compute_resources', "id = #{compute_resources(:mycompute).id}"
+    delete :destroy, { :id => compute_resources(:yourcompute).id }
+    assert_response :not_found
+  end
+
+  test "should get available images" do
+    img = Object.new
+    img.stubs(:name).returns('some_image')
+    img.stubs(:id).returns('123')
+
+    Foreman::Model::EC2.any_instance.stubs(:available_images).returns([img])
+
+    get :available_images, { :id => compute_resources(:ec2).to_param }
+    assert_response :success
+    available_images = ActiveSupport::JSON.decode(@response.body)
+    assert_not_empty available_images
+  end
+
+  test "should get available networks" do
+    network = Object.new
+    network.stubs(:name).returns('test_network')
+    network.stubs(:id).returns('my11-test35-uuid99')
+
+    Foreman::Model::Ovirt.any_instance.stubs(:available_networks).returns([network])
+
+    get :available_networks, { :id => compute_resources(:ovirt).to_param, :cluster_id => '123-456-789' }
+    assert_response :success
+    available_networks = ActiveSupport::JSON.decode(@response.body)
+    assert !available_networks.empty?
+  end
+
+  test "should get available clusters" do
+    cluster = Object.new
+    cluster.stubs(:name).returns('test_cluster')
+    cluster.stubs(:id).returns('my11-test35-uuid99')
+
+    Foreman::Model::Ovirt.any_instance.stubs(:available_clusters).returns([cluster])
+
+    get :available_clusters, { :id => compute_resources(:ovirt).to_param }
+    assert_response :success
+    available_clusters = ActiveSupport::JSON.decode(@response.body)
+    assert !available_clusters.empty?
+  end
+
+  test "should get available storage domains" do
+    storage_domain = Object.new
+    storage_domain.stubs(:name).returns('test_cluster')
+    storage_domain.stubs(:id).returns('my11-test35-uuid99')
+
+    Foreman::Model::Ovirt.any_instance.stubs(:available_storage_domains).returns([storage_domain])
+
+    get :available_storage_domains, { :id => compute_resources(:ovirt).to_param }
+    assert_response :success
+    available_storage_domains = ActiveSupport::JSON.decode(@response.body)
+    assert !available_storage_domains.empty?
+  end
+
+  context 'ec2' do 
+    setup do
+      @ec2_object = Object.new
+      @ec2_object.stubs(:name).returns('test_ec2_object')
+    end    
+
+    teardown do
+      assert_response :success
+      available_objects = ActiveSupport::JSON.decode(@response.body)
+      assert_not_empty available_objects
+    end
+
+    test "should get available flavors" do
+      @ec2_object.stubs(:id).returns('123')
+      Foreman::Model::EC2.any_instance.stubs(:available_flavors).returns([@ec2_object])

```

Comment:
This is still using stubs, you can freely call these methods now that Fog.mock! is set. It'd return an actual mocked fog response. Same goes for the rest of the stubbing done here. 

Also what I meant by a helper method was not just to use setup & teardown blocks, but making something like

```ruby
def get_available(method, compute_resource_type)
  get method, :id => compute_resources(compute_resource_type).to_param 
  assert_response :success
  assert_not_empty ActiveSupport::JSON.decode(@response.body)
end
```

Then defining your tests in terms of that so it'd be like:

```ruby
context 'ec2' do 
  get_available(:available_flavors, :ec2)
  get_available(:available_security_groups, :ec2)
  get_available(:available_zones, :ec2)
end

context 'vmware' do 
  get_available(:available_clusters, :vmware)
  get_available(:available_storage_domains, :vmware)
  get_available(:available_resource_pools, :vmware)
  get_available(:available_folders, :vmware)
end
```

Feel free to leave out the refactor of the other Compute Resource tests if you want.